src/sage/rings/polynomial/polynomial_element.pyx

```diff
@@ -2168,162 +2168,151 @@ cdef class Polynomial(CommutativePolynomial):
-                        if 2*d+2 > e:
-                            # this polynomial has no factors dividing allowed_deg_mult
-                            if allowed_deg_mult % e == 0:
-                                degree = -e
-                            break
-                        while d < allowed_deg_mult:
-                            d = d+1
-                            xq = pow(xq,q,self)
-                            if d.divides(allowed_deg_mult):
-                                break
-                        A = self.gcd(xq-x)
-                        if A != 1:
-                            self = A
-                            degree = -d
-                            break
-                        if d == allowed_deg_mult:
-                            break
-                    if degree is None:
-                        if allowed_deg_mult == 1:
-                            raise ValueError("no roots C %s" % self)
-                        xq = x
-                        d = Integer(0)
-                        while True:
-                            # now another for roots that will lie in an extension.
-                            e = self.degree()
-                            if 2*d+2 > e:
-                                # this polynomial is irreducible.
-                                degree = -e
-                                break
-                            while True:
-                                # we waste a little effort here in computing the xq again.
-                                d = d+1
-                                xq = pow(xq,q,self)
-                                if allowed_deg_mult.divides(d):
-                                    break
-                            A = self.gcd(xq-x)
-                            if A != 1:
-                                self = A
-                                degree = -d
-                                break
-            if degree == 0:
-                raise ValueError("degree should be nonzero")
-            R = self._parent
-            x = R.gen()
-            if degree > 0:
-                xq = x
-                d = 0
-                while True:
-                    e = self.degree()
-                    if 2*d > e:
-                        if degree != e:
-                            raise ValueError("no roots D %s" % self)
-                        break
-                    d = d+1
-                    xq = pow(xq,q,self)
-                    if d == degree:
-                        break
-                    A = self.gcd(xq-x)
-                    if A != 1:
-                        self = self // A
-                if d == degree:
-                    self = self.gcd(xq-x)
-                    if self.degree() == 0:
-                        raise ValueError("no roots E %s" % self)
-            else:
-                degree = -degree
+        # Currently I can't see a benefit of the degree method, so I think we should
+        # remove it
```

Comment:
I also haven't ever used this, but I'm not sure if other people have applications for it. Maybe I'm missing something, but wouldn't it be easy enough to keep the API unchanged? We simply have to divide out the $\gcd$ of $f$ with $x^{n/m}-x$ for all prime divisors $m$ of the given degree $n$ prior to running the rest of the algorithm?

Reply:
Yeah I think we can do this. The original code also allowed negative degree which did something optional before taking the absolute sign. The thing is I just don't see what the benefit is. We have some input f we want roots of and allowing degree = 1 means we find all possible linear polynomials. When the degree is larger we just find factors which may not have rational roots?

Reply:
More specifically, the old documentation says:

```py
        - ``degree`` (None or nonzero integer) -- Used for polynomials
          over finite fields.  Return a root of degree
          ``abs(degree)`` over the ground field.  If negative, also
          assumes that all factors of this polynomial are of degree
          ``abs(degree)``.  If None, returns a root of minimal degree
          contained within the given ring.
```

I don't really know what "root of degree n" means, unless what it actually means is that it first finds some polynomial A_d = prod_i^n g_i(x) with g_i(x) of degree n? In this case for a root, we seem to need d = 1 -- unless the idea is we find some polynomial of degree n over the base field which is irreducible and then find a root of this over some extension?

I wish I understood this better.

Reply:
Suppose you know a priori that the root you're looking for lies in an extension of degree $k$, but you haven't yet found a suitable irreducible polynomial to construct that extension. Then it should be faster to extract an irreducible factor from your polynomial and construct the extension with that particular polynomial (hence $x$ is the desired root) than to construct the extension using an arbitrary irreducible polynomial and running the root-finding algorithm in the extension.
One particular example where I think this may help is when constructing an ℓ‑torsion basis of an elliptic curve: We often know the required degree from point counting, but we still need to find the actual extension field and points.

Reply:
The current behavior is indeed something like that:
```
sage: x = polygen(GF(11^2))
sage: f = x^3+x^2+x+2
sage: f.any_root(degree=3)
9*z6^5 + 5*z6^4 + 5*z6^3 + 10*z6^2
```
...but it looks like it currently just runs `.extension()` on the base field to build the extension (rather than first finding an irreducible factor and making use of it), so there's essentially no point in doing this with the current implementation.

Reply:
I think a better interface for this would be `.any_irreducible_factor()` with an optional `degree=` argument. This way the user can do with the returned factor whatever they want (without having the library forcibly construct a field extension for them in case of degree $>1$), and the `.any_root()` method would be a simple wrapper.

Reply:
Ahh ok then this is indeed something which maybe could be included. Maybe I would like the parameter to be something like "extension degree" rather than degree which appears in many places... but that's a whole other thing. 

When fixing everything else from the above review I'll try my hand at this too using your idea